docs/proposals/job.md

```diff
@@ -0,0 +1,145 @@
+Job controller is similar to replication controller in that they manage pods.
+This implies they will follow the same controller framework that replication
+controllers already defined.  The biggest difference between a `Job` and a
+`ReplicationController` object is the purpose; `ReplicationController`
+ensures that a specified number of Pods are running at any one time, whereas
+`Job` is responsible for keeping the desired number of Pods to a completion of
+a task. This will be represented by the `RestartPolicy` which is required to
+always take value of `RestartPolicyNever` or `RestartOnFailure`.
+
+
+The new `Job` object will have the following content:
+
+```go
+// Job represents the configuration of a single job.
+type Job struct {
+    TypeMeta
+    ObjectMeta
+
+    // Spec is a structure defining the expected behavior of a job.
+    Spec JobSpec
+
+    // Status is a structure describing current status of a job.
+    Status JobStatus
+}
+
+// JobList is a collection of jobs.
+type JobList struct {
+    TypeMeta
+    ListMeta
+
+    Items []Job
+}
+```
+
+`JobSpec` structure is defined to contain all the information how the actual job execution
+will look like.
+
+```go
+// JobSpec describes how the job execution will look like.
+type JobSpec struct {
+
+    // TaskCount specifies the desired number of pods the job should be run with.
+    TaskCount int
+
+    // Optional duration in seconds relative to the StartTime that the job may be active
+    // before the system actively tries to terminate it; value must be positive integer
+    ActiveDeadlineSeconds *int64
+
+    // Selector is a label query over pods that should match the pod count.
+    Selector map[string]string
+
+    // Spec is the object that describes the pod that will be created when
+    // executing a job.
+    Spec PodSpec
+}
+```
+
+`JobStatus` structure is defined to contain informations about pods executing
+specified job.  The structure holds information about pods currently executing
+the job.
+
+```go
+// JobStatus represents the current state of a Job.
+type JobStatus struct {
+    // Executions holds a detailed information about each of the pods running a job.
+    Executions []JobExec
+
+    // Completions is the number of pods successfully completed their job.
+    Completions int
+
+}
+
+// JobExec represents the current state of a single execution of a Job.
+type JobExec struct {
+    // CreationTime represents time when the job execution was created
+    CreationTime util.Time
+
+    // StartTime represents time when the job execution was started
+    StartTime util.Time
+
+    // CompletionTime represents time when the job execution was completed
+    CompletionTime util.Time
+
+    // Phase represents the point in the job execution lifecycle.
+    Phase JobExecPhase
+
+    // Tag is added in labels of pod(s) created for this job execution.  It allows
+    // job object to safely group/track all pods started for one given job execution.
+    Tag util.UID
```

Comment:
What we do will set a precedent. It needs to be consistent with the approach used by other controllers. 

Are you trying to identify the pod that completed a particular task?

Tag is not an acceptable approach, since it doesn’t match what we do in any other cases. ReplicationController applies an annotation on a pod containing a serialized ObjectReference to record the creator, but it is not updated when the pod becomes managed by another RC. In Endpoints, we use an ObjectReference to represent a 1-to-1 relationship, and it is kept up to date.

The tasks/pods should be identified by a label selector, but labels are the purview of users. The user would need to at least optionally be able to choose the key of the distinguishing label. That’s what we do for rolling updates that generate a new ReplicationController. As @davidopp pointed out, we don’t currently have a way to ensure that a particular label key couldn’t be applied in other contexts, so using a fixed system-defined label key wouldn’t provide an iron-clad guarantee, anyway.

Ideally we’d choose a consistent way to prevent overlap. Originally what I had in mind was that a user would choose a label schema within a namespace and choose a unique value for at least one of the labels for each set of resources of a particular kind. We could enforce that, but it’s non-trivial to do reliably and would make simple use cases harder, since it would force users to apply more labels than they might think is necessary.

In ReplicationController, we eliminated the automatically applied label in part in order to facilitate orphaning and adoption. For Job, that’s admittedly less important. A possible approach would be to automatically apply a job=jobname label by default but allow clients to disable automatic label application.

Use of namespaces for fairly narrow purposes, such as a specific application within a specific environment (dev/test/staging/production) makes accidental overlap even less likely.

Speaking of ownership, I think we should implement existence dependencies, as proposed in #1535.